File: database/src/common/utility.rs
use sqlx::Error;
use rand::{thread_rng, Rng};
use argon2::{Argon2, PasswordHasher, PasswordVerifier, PasswordHash, password_hash::SaltString};

pub fn hash_password(password: &str) -> Result<String, argon2::password_hash::Error>
{
    let argon2 = Argon2::default();
    let salt = SaltString::generate(&mut thread_rng());
    let password_hash = argon2.hash_password(password.as_bytes(), &salt)?;
    Ok(password_hash.to_string())
}

pub fn verify_password(password: &str, password_hash: &str) -> Result<(), argon2::password_hash::Error>
{
    let parsed_hash = PasswordHash::new(password_hash).unwrap();
    Argon2::default().verify_password(password.as_bytes(), &parsed_hash)
}

pub(crate) fn verify_hash_format(password_hash: &str) -> Result<(), Error>
{
    match PasswordHash::new(password_hash) {
        Ok(_) => Ok(()),
        Err(_) => Err(Error::InvalidArgument(String::from("Invalid password hash format")))
    }
}

pub fn generate_access_key() -> Vec<u8>
{
    (0..32).map(|_| thread_rng().gen_range(0..255)).collect()
}

pub fn generate_token_string() -> String
{
    const CHARSET: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-";
    (0..32).map(|_| CHARSET[thread_rng().gen_range(0..64)] as char).collect()
}

File: grpc_api/client_python/src/bbthings_grpc/common/utility.py
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError
import random
from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP
from Crypto.Hash import SHA256


ph = PasswordHasher()

def hash_password(password: str):
    """
    Hash a password using Argon2 with a generated salt.
    Returns the encoded hash string.
    """
    return ph.hash(password)

def verify_password(password: str, password_hash: str):
    """
    Verify a password against an Argon2 hash.
    Returns True if valid; raises on failure.
    """
    try:
        ph.verify(password_hash, password)
        return True
    except VerifyMismatchError:
        return False

def encrypt_message(message: str | bytes, public_key: bytes):
    public_key = RSA.import_key(public_key)
    sha = SHA256.new()
    chipper_rsa = PKCS1_OAEP.new(public_key, sha)
    if isinstance(message, str):
        message = bytes(message, "utf-8")
    return chipper_rsa.encrypt(message)

def generate_access_key() -> bytes:
    # Generate 32 random bytes, each in range 0..254
    return bytes(random.randrange(0, 255) for _ in range(32))

def generate_token_string() -> str:
    CHARSET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-"
    # Pick 32 random characters from the charset
    return ''.join(random.choice(CHARSET) for _ in range(32))

def hex_to_bytes(s: str):
    # Length must be even
    if len(s) % 2 != 0:
        return None
    result = []
    for i in range(0, len(s), 2):
        sub = s[i:i+2]
        try:
            result.append(int(sub, 16))
        except ValueError:
            return None
    return result
